README.md
Phalcon\Ext\Mailer
==============

Mailer wrapper over SwiftMailer for Phalcon.

## Installing ##

Install composer in a common location or in your project:

    curl -s http://getcomposer.org/installer | php

Create the composer.json file as follows:

```
{
	"require": {
		"phalcon-ext/mailer": "~2.0"
	}
}
```

Run the composer installer:

```bash
php composer.phar install
```

Add in your the code

    require_once('vendor/autoload.php');

## Configure ##

**SMTP**

```php
$config = [
    'driver' 	 => 'smtp',
    'host'	 	 => 'smtp.gmail.com',
    'port'	 	 => 465,
    'encryption' => 'ssl',
    'username'   => 'example@gmail.com',
    'password'	 => 'your_password',
    'from'		 => [
    		'email' => 'example@gmail.com',
    		'name'	=> 'YOUR FROM NAME'
    	]
];
```

> **Warning**
> If you are using GMAIL and get error `...Username and Password not accepted...` or `password incorrect`, 
> please using password token ([how get token?](https://support.google.com/accounts/answer/185833)) and fix in configuration
> 
> ```php
>   ...
>
>   'username' => 'example@gmail.com',
>   'password' => 'your_password_token',
>
>   ...
> ```

**Sendmail**

```php
$config = [
    'driver' 	 => 'sendmail',
	'sendmail' 	 => '/usr/sbin/sendmail -bs',
    'from'		 => [
    	'email' => 'example@gmail.com',
    	'name'	=> 'YOUR FROM NAME'
    ]
];
```

**PHP Mail**

```php
$config = [
    'driver' 	 => 'mail',
    'from'		 => [
    	'email' => 'example@gmail.com',
    	'name'	=> 'YOUR FROM NAME'
    ]
];
```

## Example ##

### createMessage() ###

```php
$mailer = new \Phalcon\Ext\Mailer\Manager($config);

$message = $mailer->createMessage()
		->to('example_to@gmail.com', 'OPTIONAL NAME')
		->subject('Hello world!')
		->content('Hello world!');

// Set the Cc addresses of this message.
$message->cc('example_cc@gmail.com');

// Set the Bcc addresses of this message.
$message->bcc('example_bcc@gmail.com');

// Send message
$message->send();
```

### createMessageFromView() ###

> **Warning**
> If you want to use as a template engine VOLT (.volt), 
> please setup "view" service according to the official [docs](http://docs.phalconphp.com/en/latest/reference/volt.html#activating-volt) Phalcon

```php
/**
 * Global viewsDir for current instance Mailer\Manager.
 * 
 * This parameter is OPTIONAL, If it is not specified, 
 * use DI from "view" service (getViewsDir)
 */
$config['viewsDir'] = __DIR__ . '/views/email/';

$mailer = new \Phalcon\Ext\Mailer\Manager($config);

// view relative to the folder viewsDir (REQUIRED)
$viewPath = 'email/example_message';

// Set variables to views (OPTIONAL)
$params [ 
	'var1' => 'VAR VALUE 1',
	'var2' => 'VAR VALUE 2',
	...
	'varN' => 'VAR VALUE N',
];

/**
 * The local path to the folder viewsDir only this message. (OPTIONAL)
 * 
 * This parameter is OPTIONAL, If it is not specified, 
 * use global parameter "viewsDir" from configuration.
 */
$viewsDirLocal = __DIR__ . '/views/email/local/';


$message = $mailer->createMessageFromView($viewPath, $params, $viewsDirLocal)
		->to('example_to@gmail.com', 'OPTIONAL NAME')
		->subject('Hello world!');

// Set the Cc addresses of this message.
$message->cc('example_cc@gmail.com');

// Set the Bcc addresses of this message.
$message->bcc('example_bcc@gmail.com');

// Send message
$message->send();
```

## Events ##
- mailer:beforeCreateMessage
- mailer:afterCreateMessage
- mailer:beforeSend
- mailer:afterSend
- mailer:beforeAttachFile
- mailer:afterAttachFile


CHANGELOG.md
# CHANGELOG

## v3.0.0 (2021-02-24)
- Fix compatibility with Phalcon 4+ [#13](https://github.com/phalcon-ext/mailer/issues/13)

## v2.1.0 (2017-09-30)
- Fix [#7](https://github.com/phalcon-ext/mailer/issues/7)
- Add support `auth_mode` [#6](https://github.com/phalcon-ext/mailer/issues/6)
- Allow composers use Swiftmailer v6 [#8](https://github.com/phalcon-ext/mailer/pull/8)
- Change license LGPL-3.0 to MIT

## v2.0.3 (2015-12-18)
- Added method `contentAlternative()` - optionally an alternative text body 
- fix [#5](https://github.com/phalcon-ext/mailer/issues/5)

## v2.0.2 (2015-03-03)
- fix [#2](https://github.com/phalcon-ext/mailer/issues/2)

## v2.0.1 (2014-11-05)
- Changes in the private logic...
  - Added protected method `isInitSwiftMailer()`
  - Added lazy init `SwiftMailer`
  - Added check for the required of dependency injection object
  
## v2.0.0 (2014-10-16)
- rename root namespace `\Phalcon\Mailer` to `\Phalcon\Ext\Mailer`
- remove deprecated method `getMessage()` in class `\Phalcon\Ext\Mailer\Message`

## v1.1.0 (2014-10-16)
- added method `getSwiftMessage()` in class `\Phalcon\Mailer\Message` analogue for `getMessage()`
- method `getMessage()` in class `\Phalcon\Mailer\Message` is deprecated (will removed since 2.0.0)

## v1.0.2 (2014-09-17)
- fixed bug when you create a message via View (incorrect namespace)

## v1.0.1 (2014-09-15)
- fixed bug in the attachment file.
- In `\Phalcon\Mailer\Message`, creation SwiftMailer of instances moved to the DI

## v1.0.0 (2014-09-14)
- Release of component :) 


src/Mailer.php
<?php
/**
 * MailerService.php 2014-08-31 04:11
 * ----------------------------------------------
 *
 *
 * @author      Stanislav Kiryukhin <korsar.zn@gmail.com>
 * @copyright   Copyright (c) 2014, CKGroup.ru
 *
 * @version 	0.0.1
 * ----------------------------------------------
 * All Rights Reserved.
 * ----------------------------------------------
 */
namespace Phalcon\Mailer;

use Phalcon\Mvc\User\Component;
use Phalcon\Mvc\View;
use Phalcon\Config;

/**
 * Class Mailer
 * @package Phalcon\Mailer
 */
class Mailer extends Component
{
	/**
	 * @var array
	 */
	protected $config;

	/**
	 * @var mixed
	 */
	protected $transport;

	/**
	 * @var \Swift_Mailer
	 */
	protected $mailer;

	/**
	 * @var View\Simple
	 */
	protected $view;

	/**
	 * @param array $config
	 */
	public function __construct(Array $config)
	{
		$this->configure($config);
	}

	/**
	 * @return \Swift_Mailer
	 */
	public function getSwift()
	{
		return $this->mailer;
	}

	/**
	 * @param $view
	 * @param array $params
	 * @param null $viewsDir
	 *
	 * @return Message
	 */
	public function createMessageViaView($view, $params = [], $viewsDir = null)
	{
		$message = $this->createMessage();
		$message->content($this->renderView($view, $params, $viewsDir), 'text/html');


		return $message;
	}

	/**
	 * @return \Phalcon\Mailer\Message
	 */
	public function createMessage()
	{
		$eventsManager = $this->getEventsManager();

		if($eventsManager)
			$eventsManager->fire('mailer:beforeCreateMessage', $this);

		/** @var $message Message */
		$message = $this->getDI()->get('\Phalcon\Mailer\Message', [$this]);

		if(($from = $this->getConfig('from')))
			$message->from($from['email'], isset($from['name']) ? $from['name'] : null);

		if($eventsManager)
			$eventsManager->fire('mailer:afterCreateMessage', $this, [$message]);

		return $message;
	}

	/**
	 * @param $email
	 *
	 * @return string
	 */
	public function normalizeEmail($email)
	{
		if(preg_match('#[^(\x20-\x7F)]+#', $email))
		{
			list($user, $domain) = explode('@', $email);
			return $user . '@' . $this->punycode($domain);
		}
		else
			return $email;
	}


	/**
	 * Register SwiftMailer
	 */
	protected function registerSwiftMailer()
	{
		$this->mailer = new \Swift_Mailer($this->transport);
	}


	protected function registerSwiftTransport()
	{
		switch($driver = $this->getConfig('driver'))
		{
			case 'smtp':
				$this->transport = $this->registerTransportSmtp();
			break;

			case 'mail':
				$this->transport = $this->registerTransportMail();
			break;

			case 'sendmail':
				$this->transport = $this->registerTransportSendmail();
			break;

			default:
				throw new \InvalidArgumentException(sprintf('Driver-mail "%s" is not supported', $driver));
		}
	}

	/**
	 * @return \Swift_SmtpTransport
	 */
	protected function registerTransportSmtp()
	{
		$config = $this->getConfig();

		$transport = \Swift_SmtpTransport::newInstance($config['host'], $config['port']);

		if(isset($config['encryption']))
			$transport->setEncryption($config['encryption']);

		if(isset($config['username']))
		{
			$transport->setUsername($this->normalizeEmail($config['username']));
			$transport->setPassword($config['password']);
		}

		return $transport;
	}

	/**
	 * @return \Swift_SendmailTransport
	 */
	protected function registerTransportSendmail()
	{
		return \Swift_SendmailTransport::newInstance($this->getConfig('sendmail', '/usr/sbin/sendmail -bs'));
	}

	/**
	 * @return \Swift_MailTransport
	 */
	protected function registerTransportMail()
	{
		return \Swift_MailTransport::newInstance();
	}

	/**
	 * @param $config
	 */
	protected function configure(Array $config)
	{
		$this->config = $config;

		$this->registerSwiftTransport();
		$this->registerSwiftMailer();
	}

	/**
	 * @param null $key
	 * @param null $default
	 *
	 * @return mixed
	 */
	protected function getConfig($key = null, $default = null)
	{
		if($key !== null)
		{
			if(isset($this->config[$key]))
				return $this->config[$key];
			else
				return $default;
		}
		else
			return $this->config;
	}

	/**
	 * @param $viewPath
	 * @param $params
	 * @param null $viewsDir
	 *
	 * @return string
	 */
	protected function renderView($viewPath, $params, $viewsDir = null)
	{
		$view = $this->getView();

		if($viewsDir !== null)
		{
			$viewsDirOld = $view->getViewsDir();
			$view->setViewsDir($viewsDir);

			$content = $view->render($viewPath, $params);
			$view->setViewsDir($viewsDirOld);

			return $content;
		}
		else
			return $view->render($viewPath, $params);
	}

	/**
	 * @return View\Simple
	 */
	protected function getView()
	{
		if($this->view)
			return $this->view;
		else
		{
			/** @var $view \Phalcon\Mvc\View\Simple */
			$view = $this->getDI()->get('\Phalcon\View\Simple');
			$view->setViewsDir($this->getConfig('viewsDir'));

			return $this->view = $view;
		}
	}

	/**
	 * Convert UTF-8 encoded domain name to ASCII
	 *
	 * @param $str
	 *
	 * @return string
	 */
	protected function punycode($str)
	{
		if(function_exists('idn_to_ascii'))
			return idn_to_ascii($str);
		else
			return $str;
	}
} 

src/Manager.php
<?php
/**
 * Manager.php 2014-08-31 04:11
 * ----------------------------------------------
 *
 * @author      Stanislav Kiryukhin <korsar.zn@gmail.com>
 * @copyright   Copyright (c) 2014-2021
 *
 * ----------------------------------------------
 * All Rights Reserved.
 * ----------------------------------------------
 */
namespace Phalcon\Ext\Mailer;

use Phalcon\Config;
use Phalcon\Mvc\View;
use Phalcon\DiInterface;
use Phalcon\Di\Injectable;
use Phalcon\Events\ManagerInterface;
use Phalcon\Events\EventsAwareInterface;

/**
 * Class Manager
 */
class Manager extends Injectable implements EventsAwareInterface
{
    const AUTHENTICATION_MODE_CRAM_MD5  = 'CRAM-MD5';
    const AUTHENTICATION_MODE_LOGIN     = 'LOGIN';
    const AUTHENTICATION_MODE_NTLM      = 'NTLM';
    const AUTHENTICATION_MODE_PLAIN     = 'PLAIN';
    const AUTHENTICATION_MODE_OAUTH     = 'XOAUTH2';

    const ENCRYPTION_NONE = '';
    const ENCRYPTION_SSL  = 'ssl';
    const ENCRYPTION_TLS  = 'tls';

    /**
     * @var ManagerInterface
     */
    protected $eventsManager;

    /**
     * @var array
     */
    protected $config = [];

    /**
     * @var \Swift_Transport
     */
    protected $transport;

    /**
     * @var \Swift_Mailer
     */
    protected $mailer;

    /**
     * @var \Phalcon\Mvc\View\Simple
     */
    protected $view;

    /**
     * Create a new MailerManager component using $config for configuring
     *
     * @param array $config
     */
    public function __construct(array $config)
    {
        $this->configure($config);
    }

    /**
     * Returns the internal event manager
     * @return ManagerInterface
     */
    public function getEventsManager() {
        return $this->eventsManager;
    }

    /**
     * Sets the events manager
     * @return void
     */
    public function setEventsManager(ManagerInterface $eventsManager) {
        $this->eventsManager = $eventsManager;
    }

    /**
     * Create a new Message instance.
     *
     * Events:
     * - mailer:beforeCreateMessage
     * - mailer:afterCreateMessage
     *
     * @return \Phalcon\Ext\Mailer\Message
     */
    public function createMessage()
    {
        $eventsManager = $this->getEventsManager();

        if ($eventsManager) {
            $eventsManager->fire('mailer:beforeCreateMessage', $this);
        }

        /** @var $message Message */
        $message = $this->getDI()->get('\Phalcon\Ext\Mailer\Message', [$this]);

        if (($from = $this->getConfig('from'))) {
            $message->from($from['email'], isset($from['name']) ? $from['name'] : null);
        }

        if ($eventsManager) {
            $eventsManager->fire('mailer:afterCreateMessage', $this, [$message]);
        }

        return $message;
    }

    /**
     * Create a new Message instance.
     * For the body of the message uses the result of render of view
     *
     * Events:
     * - mailer:beforeCreateMessage
     * - mailer:afterCreateMessage
     *
     * @param string $view
     * @param array $params optional
     * @param null|string $viewsDir optional
     *
     * @return \Phalcon\Ext\Mailer\Message
     *
     * @see \Phalcon\Ext\Mailer\Manager::createMessage()
     */
    public function createMessageFromView($view, $params = [], $viewsDir = null)
    {
        $message = $this->createMessage();
        $message->content($this->renderView($view, $params, $viewsDir), $message::CONTENT_TYPE_HTML);

        return $message;
    }

    /**
     * Return a {@link \Swift_Mailer} instance
     *
     * @return \Swift_Mailer
     */
    public function getSwift()
    {
        if (!$this->isInitSwiftMailer()) {
            $this->registerSwiftMailer();
        }

        return $this->mailer;
    }

    /**
     * Normalize IDN domains.
     *
     * @param $email
     *
     * @return string
     *
     * @see \Phalcon\Ext\Mailer\Manager::punycode()
     */
    public function normalizeEmail($email)
    {
        if (preg_match('#[^(\x20-\x7F)]+#', $email)) {

            list($user, $domain) = explode('@', $email);

            return $user . '@' . $this->punycode($domain);

        } else {
            return $email;
        }
    }

    /**
     * Configure MailerManager class
     *
     * @param array $config
     *
     * @see \Phalcon\Ext\Mailer\Manager::registerSwiftTransport()
     * @see \Phalcon\Ext\Mailer\Manager::registerSwiftMailer()
     */
    protected function configure(array $config)
    {
        $this->config = $config;
    }

    /**
     * Create a new Driver-mail of SwiftTransport instance.
     *
     * Supported driver-mail:
     * - smtp
     * - sendmail
     * - mail
     *
     */
    protected function registerSwiftTransport()
    {
        switch ($driver = $this->getConfig('driver')) {
            case 'smtp':
                $this->transport = $this->registerTransportSmtp();
                break;

            case 'mail':
                $this->transport = $this->registerTransportMail();
                break;

            case 'sendmail':
                $this->transport = $this->registerTransportSendmail();
                break;

            default:
                throw new \InvalidArgumentException(sprintf('Driver-mail "%s" is not supported', $driver));
        }
    }

    /**
     * Create a new SmtpTransport instance.
     *
     * @return \Swift_SmtpTransport
     *
     * @see \Swift_SmtpTransport
     */
    protected function registerTransportSmtp()
    {
        $config = $this->getConfig();

        /** @var $transport \Swift_SmtpTransport: */
        $transport = $this->getDI()->get('\Swift_SmtpTransport');

        if (isset($config['encryption'])) {
            $transport->setEncryption($config['encryption']);
        }

        if (isset($config['host'])) {
            $transport->setHost($config['host']);
        } else {
            $transport->setHost('localhost');
        }

        if (isset($config['port'])) {
            $transport->setPort($config['port']);
        } else {
            switch ($transport->getEncryption()) {
                case static::ENCRYPTION_SSL:
                    $transport->setPort(465);
                    break;

                case static::ENCRYPTION_TLS:
                    $transport->setPort(587);
                    break;

                default:
                    $transport->setPort(25);
            }
        }

        if (isset($config['auth_mode'])) {
            $transport->setAuthMode($config['auth_mode']);
        }

        if (isset($config['username'])) {
            $transport->setUsername($this->normalizeEmail($config['username']));
        }

        if (isset($config['password'])) {
            $transport->setPassword($config['password']);
        }

        return $transport;
    }

    /**
     * Get option config or the entire array of config, if the parameter $key is not specified.
     *
     * @param null $key
     * @param null $default
     *
     * @return string|array
     */
    protected function getConfig($key = null, $default = null)
    {
        if ($key !== null) {
            if (isset($this->config[$key])) {
                return $this->config[$key];
            } else {
                return $default;
            }

        } else {
            return $this->config;
        }
    }

    /**
     * Convert UTF-8 encoded domain name to ASCII
     *
     * @param $str
     *
     * @return string
     */
    protected function punycode($str)
    {
        if (function_exists('idn_to_ascii')) {
            return idn_to_ascii($str);
        } else {
            return $str;
        }
    }

    /**
     * Create a new MailTransport instance.
     *
     * @return \Swift_MailTransport
     *
     * @see \Swift_MailTransport
     */
    protected function registerTransportMail()
    {
        return $this->getDI()->get('\Swift_MailTransport');
    }

    /**
     * Create a new SendmailTransport instance.
     *
     * @return \Swift_SendmailTransport
     *
     * @see \Swift_SendmailTransport
     */
    protected function registerTransportSendmail()
    {
        /** @var $transport \Swift_SendmailTransport */
        $transport = $this->getDI()->get('\Swift_SendmailTransport')
            ->setCommand($this->getConfig('sendmail', '/usr/sbin/sendmail -bs'));

        return $transport;
    }

    /**
     * Register SwiftMailer
     *
     * @see \Swift_Mailer
     */
    protected function registerSwiftMailer()
    {
        $this->registerSwiftTransport();
        $this->mailer = $this->getDI()->get('\Swift_Mailer', [$this->transport]);
    }

    /**
     * Renders a view
     *
     * @param $viewPath
     * @param $params
     * @param null $viewsDir
     *
     * @return string
     */
    protected function renderView($viewPath, $params, $viewsDir = null)
    {
        $view = $this->getView();

        if ($viewsDir !== null) {
            $view = clone $view;
            $view->setViewsDir($viewsDir);
        }

        if (!$view->getViewsDir()) {
            throw new \RuntimeException('Please set config property "viewsDir"');
        }

        return $view->render($viewPath, $params);
    }

    /**
     * Return a {@link \Phalcon\Mvc\View\Simple} instance
     *
     * @return \Phalcon\Mvc\View\Simple
     */
    protected function getView()
    {
        if ($this->view) {
            return $this->view;
        } else {

            /** @var $viewApp \Phalcon\Mvc\View */
            $viewApp = $this->getDI()->get('view');

            if (!($viewsDir = $this->getConfig('viewsDir'))) {
                $viewsDir = $viewApp->getViewsDir();
            }

            /** @var $view \Phalcon\Mvc\View\Simple */
            $view = $this->getDI()->get('\Phalcon\Mvc\View\Simple');
            $view->setViewsDir($viewsDir);

            if ($engines = $viewApp->getRegisteredEngines()) {
                $view->registerEngines($engines);
            }

            return $this->view = $view;
        }
    }

    /**
     * Check init SwiftMailer
     *
     * @return bool
     */
    protected function isInitSwiftMailer()
    {
        return $this->mailer && $this->transport;
    }

    /**
     * Returns the internal dependency injector
     *
     * @return \Phalcon\DiInterface
     */
    public function getDI()
    {
        if (!($di = parent::getDI()) && !($di instanceof DiInterface)) {
            throw new \RuntimeException('A dependency injection object is required to access internal services');
        }

        return $di;
    }
}
